### ISSUE TITLE
Java Guestbook sample should have a parent pom.xml

### ISSUE THREAD

**Issue by etanshaul** [GoogleCloudPlatform/cloud-code-samples#24]
currently the guestbook sample has individual pom.xml's under the `backend` and `frontend` services.

This doesn't work well with the IntelliJ multi-module model where the project is expected to have a root pom.xml. With the current setup, the user is unable to sync the project with Maven unless the "sub" poms are individually loaded (by browsing to them in the Maven window).

**Comment by DanSanche** [comment]
Do we need to sync the project as a whole with maven? Won't users of these samples be mostly interacting with the container images rather than the local java environment?

Adding a parent pom that connects to all the sub-components seems to be breaking the spirit of microservices, and it could add more confusion to the project (it makes it seem like the services are more tightly coupled than they are)

**Comment by etanshaul** [comment]
hey @DanSanche 

Here is what happens, as is, with the project.

1) open the root `java-guestbook` app in IntelliJ. Notice the maven modules are not detected:
![image](https://user-images.githubusercontent.com/1735744/55485021-99f89900-55f7-11e9-9d49-4a22f29ea605.png)

2) look at the module setup under `Project Structure`. Notice the `frontend` and `backend` modules are not present. Most importantly, no source roots are found. This means that the java files will not be compiled and treated sources. This is because there is no root pom.xml. IntelliJ does not automatically pick up the submodules' poms.
![image](https://user-images.githubusercontent.com/1735744/55485100-bac0ee80-55f7-11e9-9cdf-b5474fafc375.png)
![image](https://user-images.githubusercontent.com/1735744/55485199-f5c32200-55f7-11e9-813e-1909d0e881a5.png)

Note: At this point, you _could_ deploy the application to Kubernetes using the Run Configurations and it will work. This is independent of the fact that the IDE is not aware of the project structure. But any attempt to add or modify code would be like editing a plain text file.
![image](https://user-images.githubusercontent.com/1735744/55485156-dd530780-55f7-11e9-853c-5dbcba1479e1.png)

3) To fix this, you can manually add each submodule pom to the maven structure of the project:
![image](https://user-images.githubusercontent.com/1735744/55485282-273bed80-55f8-11e9-8ad7-4b49f786036e.png)
![image](https://user-images.githubusercontent.com/1735744/55485316-37ec6380-55f8-11e9-8fdd-8c2d83ebfa5d.png)
![image](https://user-images.githubusercontent.com/1735744/55485329-40449e80-55f8-11e9-91c1-ceb3a468fa36.png)


> Adding a parent pom that connects to all the sub-components seems to be breaking the spirit of microservices

While I see your point here, the sample in its current state has a skaffold configuration that runs _both microservices_ at the same time.  Presumably the user would want to open the root guestbook app and have a view of both services in the IDE. If the user only wanted to deal with a single microservice, they could load only the subproject into their IDE and things would compile correctly - however, we currently do not have the ability to load a `skaffold.yaml` that lives outside of the project root (in this case the `skaffold.yaml` would live in a parent directory outside of the project).

I guess this ultimately depends on what we are trying to present the user: a multimodule project with a couple services, or independent microservices developed independently.

Maybe @rayt as some input on a canonical setup for a simple 2 service guestbook app with a `frontend` and a `backend`?

**Comment by etanshaul** [comment]
@DanSanche could you give @saturnism (Ray) access to this repo if he doesn't aleady? He may have some valuable input on the java projects.

**Comment by simonz130** [comment]
Etan, the repo is public, @saturnism is welcome to comment and contribute.

**Comment by DanSanche** [comment]
That's surprising that IntelliJ won't catch Java errors without the root pom file. Since it seems like IntelliJ requires rigid project structure, would it support even support an app made up of many microservices written in different languages? Or would that break the IDEs expectations like this? I'll probably have to experiment with IntelliJ before I can comment more

In general, adding a root pom that references the others wouldn't be the end of the world, but it could be confusing to new users. I'd prefer if all language-specific files were contained completely in the directory for the service

**Comment by etanshaul** [comment]
> That's surprising that IntelliJ won't catch Java errors without the root pom file

Yeah thats mainly coming from the module screenshot that I showed above. There are no source roots defined, therefore IntelliJ is not treating any of the files as Java files.

Adding a root pom would work. Although this is a bit tricky given the Dockerfiles in each submodule which would need access to the parent pom (which is outside of the Docker context itself). 

The other option, if adding a parent pom, is to update Skaffold to use Jib and get rid of the Dockerfiles.

**Comment by simonz130** [comment]
I am not comfortable with using Jib for Java and docker files for all the rest. Developers who might be interested in creating microservices in different languages will find it a bit confusing.
Also, parent pom sounds like a stop gap to address IntelliJ issues. 

Let's see if Ray has ideas on this.

**Comment by etanshaul** [comment]
That's fine. In the meantime, do you want to tackle adding a parent pom?

**Comment by saturnism** [comment]
+1 to have a parent pom. the original (simpler) was here: https://github.com/saturnism/helloworld-java-springboot.

We should start off simple w/o little dependency as possible.

**Comment by saturnism** [comment]
i noticed that the current sample uses a Dockerfile. For Java, we should use Jib, and also a newer version of OpenJDK version (adoptopenjdk:openjdk8)

**Comment by etanshaul** [comment]
@simonz130 given what Ray suggested, are you ok making these changes? I'll still go ahead and instruct tech writer to use the guestbook sample in the docs since at least the parent pom issue will get resolved. I was considering falling back to helloworld depending on how quickly we can resolve this.

**Comment by simonz130** [comment]
We wanted the samples to be consistent across different languages. This include working with docker files. How Jib plays with Skaffold? We never tested it and I'm not comfortable to make this change in the last minute.

As to root pom - we might have to do it as I don't see anything better we can do for now.
@DanSanche - are you OK with adding root pom?

**Comment by DanSanche** [comment]
Yeah, it seems like we probably need to add the pom to support IntelliJ at this point. I threw togther a PR with simple parent pom here https://github.com/GoogleCloudPlatform/cloud-code-samples/pull/35

I'm not too familiar with intelliJ though, so idk if this pom will work automatically with the intended work flow. Someone more familiar should try it out and give feedback/fixes. (I seemed to get a couple "cannot resolve symbol" issues, but I think that's because I didn't set up the IDE properly last time I used it)
